File: ui/src/api/backendSetup.ts
/**
 * Backend setup API - works without Python bridge
 *
 * These functions handle backend detection, installation, and configuration
 * before the Python bridge is running.
 */

import { invoke } from '@tauri-apps/api/core';
import type {
  CommandResponse,
  BackendType,
  BackendConfig,
  GlobalBackendConfig,
  PrerequisiteStatus,
} from '../types';

/**
 * Detect system prerequisites for backend installation.
 *
 * Detects Python, DevFlow package, Docker, and WSL2 availability.
 */
export async function detectPrerequisites(): Promise<PrerequisiteStatus> {
  const response = await invoke<CommandResponse<PrerequisiteStatus>>('detect_prerequisites');
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to detect prerequisites');
  }
  return response.data;
}

/**
 * Get the current backend configuration.
 *
 * Returns the global backend config from ~/.devflow/backend.json.
 */
export async function getBackendConfig(): Promise<GlobalBackendConfig> {
  const response = await invoke<CommandResponse<GlobalBackendConfig>>('get_backend_config');
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get backend config');
  }
  return response.data;
}

/**
 * Save the backend configuration.
 *
 * Saves to ~/.devflow/backend.json.
 */
export async function saveBackendConfig(config: BackendConfig): Promise<void> {
  const response = await invoke<CommandResponse<void>>('save_backend_config', { config });
  if (!response.success) {
    throw new Error(response.error || 'Failed to save backend config');
  }
}

/**
 * Install the backend based on type.
 *
 * For local_python: Runs pip install devflow
 * For docker: Pulls image and starts container
 * For wsl2: Installs devflow in WSL2 and starts service
 * For remote: Verifies connection to remote backend
 */
export async function installBackend(
  backendType: BackendType,
  config?: Partial<BackendConfig>
): Promise<string> {
  const response = await invoke<CommandResponse<string>>('install_backend', {
    backendType,
    config: config || null,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to install backend');
  }
  return response.data;
}

/**
 * Start the backend service (Docker container or WSL2 service).
 */
export async function startBackendService(config: BackendConfig): Promise<void> {
  const response = await invoke<CommandResponse<void>>('start_backend_service', { config });
  if (!response.success) {
    throw new Error(response.error || 'Failed to start backend service');
  }
}

/**
 * Stop the backend service (Docker container or WSL2 service).
 */
export async function stopBackendService(config: BackendConfig): Promise<void> {
  const response = await invoke<CommandResponse<void>>('stop_backend_service', { config });
  if (!response.success) {
    throw new Error(response.error || 'Failed to stop backend service');
  }
}

/**
 * Test connection to the backend.
 *
 * Returns true if the backend is accessible and responding.
 */
export async function testBackendConnection(config: BackendConfig): Promise<boolean> {
  const response = await invoke<CommandResponse<boolean>>('test_backend_connection', { config });
  if (!response.success) {
    throw new Error(response.error || 'Failed to test backend connection');
  }
  return response.data ?? false;
}

/**
 * Configure the bridge manager with the backend config and start it.
 */
export async function startBridgeWithConfig(config: BackendConfig): Promise<void> {
  const response = await invoke<CommandResponse<void>>('start_bridge_with_config', { config });
  if (!response.success) {
    throw new Error(response.error || 'Failed to start bridge with config');
  }
}

/**
 * Get recommended backend type based on prerequisites.
 */
export async function getRecommendedBackend(): Promise<BackendType> {
  const response = await invoke<CommandResponse<BackendType>>('get_recommended_backend');
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get recommended backend');
  }
  return response.data;
}

/**
 * Create a default BackendConfig for a given type.
 */
export function createDefaultConfig(backendType: BackendType): BackendConfig {
  switch (backendType) {
    case 'local_python':
      return {
        backend_type: 'local_python',
        python_path: null,
        container_name: null,
        wsl_distro: null,
        remote_host: null,
        remote_port: null,
        auto_start: true,
      };
    case 'docker':
      return {
        backend_type: 'docker',
        python_path: null,
        container_name: 'devflow-backend',
        wsl_distro: null,
        remote_host: '127.0.0.1',
        remote_port: 9876,
        auto_start: true,
      };
    case 'wsl2':
      return {
        backend_type: 'wsl2',
        python_path: null,
        container_name: null,
        wsl_distro: 'Ubuntu',
        remote_host: '127.0.0.1',
        remote_port: 9876,
        auto_start: true,
      };
    case 'remote':
      return {
        backend_type: 'remote',
        python_path: null,
        container_name: null,
        wsl_distro: null,
        remote_host: 'localhost',
        remote_port: 9876,
        auto_start: false,
      };
  }
}

/**
 * Get a human-readable name for a backend type.
 */
export function getBackendTypeName(backendType: BackendType): string {
  switch (backendType) {
    case 'local_python':
      return 'Local Python';
    case 'docker':
      return 'Docker Container';
    case 'wsl2':
      return 'WSL2 Service';
    case 'remote':
      return 'Remote Server';
  }
}

/**
 * Get a description for a backend type.
 */
export function getBackendTypeDescription(backendType: BackendType): string {
  switch (backendType) {
    case 'local_python':
      return 'Run DevFlow directly using your local Python installation. Recommended for most users.';
    case 'docker':
      return 'Run DevFlow in a Docker container. Isolated environment, easy to manage.';
    case 'wsl2':
      return 'Run DevFlow as a service in Windows Subsystem for Linux 2. Best for Windows users.';
    case 'remote':
      return 'Connect to a DevFlow instance running on a remote server.';
  }
}

File: ui/src-tauri/src/commands/backend_setup.rs
//! Backend setup commands - work without Python bridge.
//!
//! These commands handle backend detection, installation, and configuration
//! before the Python bridge is running.

use super::CommandResponse;
use crate::backend::{
    check_devflow_installed, check_docker_container, detect_all_prerequisites, detect_docker,
    detect_python, detect_wsl, install_devflow_local, install_devflow_wsl, pull_docker_image,
    remove_docker_container, start_docker_container, start_wsl_service, stop_docker_container,
    stop_wsl_service, test_devflow_connection, BackendConfig, BackendType, GlobalBackendConfig,
    PrerequisiteStatus,
};
use crate::bridge::{BridgeManager, ConnectionMode};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Arc;
use tauri::State;

/// Response for prerequisite detection.
#[derive(Debug, Serialize, Deserialize)]
pub struct PrerequisiteResponse {
    pub python_available: bool,
    pub python_version: Option<String>,
    pub python_path: Option<String>,
    pub devflow_installed: bool,
    pub devflow_version: Option<String>,
    pub docker_available: bool,
    pub docker_running: bool,
    pub docker_version: Option<String>,
    pub wsl_available: bool,
    pub wsl_distros: Vec<String>,
}

impl From<PrerequisiteStatus> for PrerequisiteResponse {
    fn from(status: PrerequisiteStatus) -> Self {
        Self {
            python_available: status.python_available,
            python_version: status.python_version,
            python_path: status.python_path.map(|p| p.to_string_lossy().to_string()),
            devflow_installed: status.devflow_installed,
            devflow_version: status.devflow_version,
            docker_available: status.docker_available,
            docker_running: status.docker_running,
            docker_version: status.docker_version,
            wsl_available: status.wsl_available,
            wsl_distros: status.wsl_distros,
        }
    }
}

/// Detect system prerequisites for backend installation.
///
/// This command runs without the Python bridge and detects:
/// - Python installation
/// - DevFlow package installation
/// - Docker availability
/// - WSL2 availability (Windows only)
#[tauri::command]
pub fn detect_prerequisites() -> CommandResponse<PrerequisiteResponse> {
    log::info!("Detecting system prerequisites");
    let status = detect_all_prerequisites();
    CommandResponse::ok(PrerequisiteResponse::from(status))
}

/// Get the current backend configuration.
///
/// Returns the global backend config from ~/.devflow/backend.json.
#[tauri::command]
pub fn get_backend_config() -> CommandResponse<GlobalBackendConfig> {
    log::info!("Loading backend configuration");
    let config = GlobalBackendConfig::load();
    CommandResponse::ok(config)
}

/// Save the backend configuration.
///
/// Saves to ~/.devflow/backend.json.
#[tauri::command]
pub fn save_backend_config(config: BackendConfig) -> CommandResponse<()> {
    log::info!("Saving backend configuration: {:?}", config.backend_type);

    let mut global_config = GlobalBackendConfig::load();
    global_config.set_configured(config);

    match global_config.save() {
        Ok(()) => CommandResponse::ok(()),
        Err(e) => CommandResponse::err(e),
    }
}

/// Install the backend based on type.
#[tauri::command]
pub fn install_backend(backend_type: BackendType, config: Option<Value>) -> CommandResponse<String> {
    log::info!("Installing backend: {:?}", backend_type);

    match backend_type {
        BackendType::LocalPython => {
            // Get python path from config if provided
            let python_path = config
                .and_then(|c| c.get("python_path")?.as_str().map(|s| s.into()));

            let result = install_devflow_local(python_path.as_ref());
            if result.success {
                CommandResponse::ok(result.message)
            } else {
                CommandResponse::err(result.message)
            }
        }
        BackendType::Docker => {
            // Pull the image
            let pull_result = pull_docker_image();
            if !pull_result.success {
                return CommandResponse::err(pull_result.message);
            }

            // Start the container
            let container_name = config
                .as_ref()
                .and_then(|c| c.get("container_name")?.as_str())
                .unwrap_or("devflow-backend");
            let port = config
                .as_ref()
                .and_then(|c| c.get("port")?.as_u64())
                .unwrap_or(9876) as u16;

            let start_result = start_docker_container(container_name, port);
            if start_result.success {
                CommandResponse::ok(start_result.message)
            } else {
                CommandResponse::err(start_result.message)
            }
        }
        BackendType::Wsl2 => {
            let distro = config
                .as_ref()
                .and_then(|c| c.get("wsl_distro")?.as_str())
                .unwrap_or("Ubuntu");
            let port = config
                .as_ref()
                .and_then(|c| c.get("port")?.as_u64())
                .unwrap_or(9876) as u16;

            // Install devflow in WSL
            let install_result = install_devflow_wsl(distro);
            if !install_result.success {
                return CommandResponse::err(install_result.message);
            }

            // Start the service
            let start_result = start_wsl_service(distro, port);
            if start_result.success {
                CommandResponse::ok(start_result.message)
            } else {
                CommandResponse::err(start_result.message)
            }
        }
        BackendType::Remote => {
            // Remote backend doesn't need installation, just verify connection
            let host = config
                .as_ref()
                .and_then(|c| c.get("remote_host")?.as_str())
                .unwrap_or("127.0.0.1");
            let port = config
                .as_ref()
                .and_then(|c| c.get("remote_port")?.as_u64())
                .unwrap_or(9876) as u16;

            if test_devflow_connection(host, port) {
                CommandResponse::ok("Remote backend is accessible".to_string())
            } else {
                CommandResponse::err(format!(
                    "Cannot connect to remote backend at {}:{}",
                    host, port
                ))
            }
        }
    }
}

/// Start the backend service (Docker container or WSL2 service).
#[tauri::command]
pub fn start_backend_service(config: BackendConfig) -> CommandResponse<()> {
    log::info!("Starting backend service: {:?}", config.backend_type);

    match config.backend_type {
        BackendType::LocalPython => {
            // Local Python doesn't have a separate service - the bridge handles it
            CommandResponse::ok(())
        }
        BackendType::Docker => {
            let container_name = config.container_name.as_deref().unwrap_or("devflow-backend");
            let port = config.tcp_port();
            let result = start_docker_container(container_name, port);
            if result.success {
                CommandResponse::ok(())
            } else {
                CommandResponse::err(result.message)
            }
        }
        BackendType::Wsl2 => {
            let distro = config.wsl_distro.as_deref().unwrap_or("Ubuntu");
            let port = config.tcp_port();
            let result = start_wsl_service(distro, port);
            if result.success {
                CommandResponse::ok(())
            } else {
                CommandResponse::err(result.message)
            }
        }
        BackendType::Remote => {
            // Remote backend is managed externally
            CommandResponse::ok(())
        }
    }
}

/// Stop the backend service (Docker container or WSL2 service).
#[tauri::command]
pub fn stop_backend_service(config: BackendConfig) -> CommandResponse<()> {
    log::info!("Stopping backend service: {:?}", config.backend_type);

    match config.backend_type {
        BackendType::LocalPython => {
            // Local Python is handled by the bridge
            CommandResponse::ok(())
        }
        BackendType::Docker => {
            let container_name = config.container_name.as_deref().unwrap_or("devflow-backend");
            let result = stop_docker_container(container_name);
            if result.success {
                CommandResponse::ok(())
            } else {
                CommandResponse::err(result.message)
            }
        }
        BackendType::Wsl2 => {
            let distro = config.wsl_distro.as_deref().unwrap_or("Ubuntu");
            let port = config.tcp_port();
            let result = stop_wsl_service(distro, port);
            if result.success {
                CommandResponse::ok(())
            } else {
                CommandResponse::err(result.message)
            }
        }
        BackendType::Remote => {
            // Remote backend is managed externally
            CommandResponse::ok(())
        }
    }
}

/// Test connection to the backend.
#[tauri::command]
pub fn test_backend_connection(config: BackendConfig) -> CommandResponse<bool> {
    log::info!("Testing backend connection: {:?}", config.backend_type);

    match config.backend_type {
        BackendType::LocalPython => {
            // For local Python, we check if devflow is installed
            let python_path = config.python_path.as_ref();
            let (installed, _version) = check_devflow_installed(python_path);
            CommandResponse::ok(installed)
        }
        BackendType::Docker => {
            // Check if container is running and responsive
            let container_name = config.container_name.as_deref().unwrap_or("devflow-backend");
            let (_exists, running) = check_docker_container(container_name);
            if !running {
                return CommandResponse::ok(false);
            }
            // Test the TCP connection
            let connected = test_devflow_connection(&config.tcp_host(), config.tcp_port());
            CommandResponse::ok(connected)
        }
        BackendType::Wsl2 | BackendType::Remote => {
            // Test TCP connection
            let connected = test_devflow_connection(&config.tcp_host(), config.tcp_port());
            CommandResponse::ok(connected)
        }
    }
}

/// Configure the bridge manager with the backend config and start it.
#[tauri::command]
pub async fn start_bridge_with_config(
    bridge: State<'_, Arc<BridgeManager>>,
    config: BackendConfig,
) -> Result<CommandResponse<()>, ()> {
    log::info!("Starting bridge with config: {:?}", config.backend_type);

    match config.backend_type {
        BackendType::LocalPython => {
            bridge.set_mode(ConnectionMode::Subprocess);
            if let Some(ref python_path) = config.python_path {
                bridge.set_python_path(python_path.clone());
            }
            // Start in subprocess mode
            match bridge.start("bridge.main", None) {
                Ok(()) => Ok(CommandResponse::ok(())),
                Err(e) => Ok(CommandResponse::err(format!("{}", e))),
            }
        }
        BackendType::Docker | BackendType::Wsl2 | BackendType::Remote => {
            bridge.set_mode(ConnectionMode::Tcp);
            bridge.set_tcp_config(config.tcp_host(), config.tcp_port());
            // Start in TCP mode
            match bridge.start("", None) {
                Ok(()) => Ok(CommandResponse::ok(())),
                Err(e) => Ok(CommandResponse::err(format!("{}", e))),
            }
        }
    }
}

/// Get recommended backend type based on prerequisites.
#[tauri::command]
pub fn get_recommended_backend() -> CommandResponse<BackendType> {
    let status = detect_all_prerequisites();

    // Priority: LocalPython (if devflow installed) > Docker (if running) > WSL2 > LocalPython (if python available)
    if status.devflow_installed {
        return CommandResponse::ok(BackendType::LocalPython);
    }

    if status.docker_available && status.docker_running {
        return CommandResponse::ok(BackendType::Docker);
    }

    if status.wsl_available && !status.wsl_distros.is_empty() {
        return CommandResponse::ok(BackendType::Wsl2);
    }

    if status.python_available {
        return CommandResponse::ok(BackendType::LocalPython);
    }

    // Default to Docker if nothing else is available
    CommandResponse::ok(BackendType::Docker)
}
